components/raftstore/src/store/peer.rs

```diff
@@ -4502,27 +4513,32 @@ where
         force: bool,
     ) {
         if let Some(unsafe_recovery_state) = &self.unsafe_recovery_state {
-            let UnsafeRecoveryState::WaitApply {
+            if let UnsafeRecoveryState::WaitApply {
                 target_index,
                 task_counter,
-            } = unsafe_recovery_state;
-            if self.raft_group.raft.raft_log.applied >= *target_index || force {
-                info!(
-                    "unsafe recovery finish wait apply";
-                    "region_id" => self.region().get_id(),
-                    "peer_id" => self.peer_id(),
-                    "target_index" => target_index,
-                    "applied" =>  self.raft_group.raft.raft_log.applied,
-                    "force" => force,
-                    "counter" =>  task_counter.load(Ordering::SeqCst)
-                );
-                if task_counter.fetch_sub(1, Ordering::SeqCst) == 1 {
-                    if let Err(e) = ctx.router.send_control(StoreMsg::UnsafeRecoveryReport) {
-                        error!("fail to send detailed report after recovery tasks finished"; "err" => ?e);
+            } = unsafe_recovery_state
+            {
+                if self.raft_group.raft.raft_log.applied >= *target_index || force {
```

Comment:
in the force leader state, wait apply should wait until the commit index == last index